{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": []
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "source": [
        "# RNA-seq Pipeline practice\n",
        "\n",
        "We are going to use real RNA-seq data from the following publication: https://www.cell.com/heliyon/fulltext/S2405-8440(24)00413-4. GEO accession number: GSE250471\n",
        "\n",
        "This data comes from a human macrophages cell line (U937). The data is available on GEO (count matrix) and also SRA (FASTQ files).\n",
        "However, To reduce the computing time, we will be working only with reads that align to the chromosome 22.\n",
        "\n"
      ],
      "metadata": {
        "id": "yw54y_OYRkiI"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "1. First we are going to download and install all the necessary softwares and data to perform the analysis:\n",
        "\n",
        "        a) FASTQC (quality control)\n",
        "\n",
        "        b) Trimmomatic (trimming)\n",
        "\n",
        "        b) STAR (alignment)\n",
        "\n",
        "        c) samtools (BAM filtering)\n",
        "\n",
        "        d) FeatureCounts (generating count matrix)\n",
        "\n",
        "        e) GTF file with genome annotations and chr22 Hg38 reference fasta file\n",
        "\n",
        "2. We will perform quality control of raw FASTQ files using FASTQC\n",
        "\n",
        "3. Trimming using Trimmomatic and post-trimming QC\n",
        "\n",
        "4. Alignment using STAR\n",
        "\n",
        "5. Indexing and filtering the BAM file\n",
        "\n",
        "6. Generating count matrix with FeatureCounts\n",
        "\n"
      ],
      "metadata": {
        "id": "vsQMJqVuSJgU"
      }
    },
    {
      "cell_type": "code",
      "execution_count": 1,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "RSsuEUqFGcgi",
        "outputId": "a2649042-2be1-46f8-ed81-eb8711c26c5f"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Cloning into 'STAR'...\n",
            "remote: Enumerating objects: 11142, done.\u001b[K\n",
            "remote: Counting objects: 100% (409/409), done.\u001b[K\n",
            "remote: Compressing objects: 100% (53/53), done.\u001b[K\n",
            "remote: Total 11142 (delta 376), reused 356 (delta 356), pack-reused 10733 (from 2)\u001b[K\n",
            "Receiving objects: 100% (11142/11142), 529.00 MiB | 18.17 MiB/s, done.\n",
            "Resolving deltas: 100% (7867/7867), done.\n",
            "2.7.11b\n"
          ]
        }
      ],
      "source": [
        "!git clone https://github.com/alexdobin/STAR.git\n",
        "!./STAR/bin/Linux_x86_64_static/STAR --version"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Download FastQC\n",
        "!wget https://www.bioinformatics.babraham.ac.uk/projects/fastqc/fastqc_v0.12.1.zip\n",
        "!unzip fastqc_v0.12.1.zip\n",
        "!chmod +x FastQC/fastqc\n",
        "\n",
        "# Run FastQC on a FASTQ file\n",
        "!FastQC/fastqc /content/test.fastq -o /content/"
      ],
      "metadata": {
        "id": "xewGDCwVSswq",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "69654497-339c-4adb-b63f-a4abbdc6df15"
      },
      "execution_count": 2,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "--2025-09-30 19:30:09--  https://www.bioinformatics.babraham.ac.uk/projects/fastqc/fastqc_v0.12.1.zip\n",
            "Resolving www.bioinformatics.babraham.ac.uk (www.bioinformatics.babraham.ac.uk)... 149.155.133.4\n",
            "Connecting to www.bioinformatics.babraham.ac.uk (www.bioinformatics.babraham.ac.uk)|149.155.133.4|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 11709692 (11M) [application/zip]\n",
            "Saving to: ‘fastqc_v0.12.1.zip’\n",
            "\n",
            "fastqc_v0.12.1.zip  100%[===================>]  11.17M  8.42MB/s    in 1.3s    \n",
            "\n",
            "2025-09-30 19:30:11 (8.42 MB/s) - ‘fastqc_v0.12.1.zip’ saved [11709692/11709692]\n",
            "\n",
            "Archive:  fastqc_v0.12.1.zip\n",
            "   creating: FastQC/\n",
            "  inflating: FastQC/htsjdk.jar       \n",
            "  inflating: FastQC/README.md        \n",
            "  inflating: FastQC/INSTALL.txt      \n",
            "  inflating: FastQC/LICENSE          \n",
            "   creating: FastQC/net/\n",
            "   creating: FastQC/net/sourceforge/\n",
            "   creating: FastQC/net/sourceforge/iharder/\n",
            "   creating: FastQC/net/sourceforge/iharder/base64/\n",
            "  inflating: FastQC/net/sourceforge/iharder/base64/Base64.class  \n",
            "  inflating: FastQC/net/sourceforge/iharder/base64/Base64$InputStream.class  \n",
            "  inflating: FastQC/net/sourceforge/iharder/base64/Base64$1.class  \n",
            "  inflating: FastQC/net/sourceforge/iharder/base64/Base64$OutputStream.class  \n",
            "  inflating: FastQC/LICENSE_JHDF5.txt  \n",
            "   creating: FastQC/Templates/\n",
            "   creating: FastQC/Templates/Icons/\n",
            " extracting: FastQC/Templates/Icons/tick.png  \n",
            " extracting: FastQC/Templates/Icons/error.png  \n",
            " extracting: FastQC/Templates/Icons/fastqc_icon.png  \n",
            " extracting: FastQC/Templates/Icons/warning.png  \n",
            "  inflating: FastQC/Templates/fastqc2fo.xsl  \n",
            "  inflating: FastQC/Templates/header_template.html  \n",
            "  inflating: FastQC/README.txt       \n",
            "  inflating: FastQC/RELEASE_NOTES.txt  \n",
            "   creating: FastQC/Configuration/\n",
            "  inflating: FastQC/Configuration/contaminant_list.txt  \n",
            "  inflating: FastQC/Configuration/adapter_list.txt  \n",
            "  inflating: FastQC/Configuration/limits.txt  \n",
            "  inflating: FastQC/LICENSE.txt      \n",
            "  inflating: FastQC/fastqc           \n",
            "  inflating: FastQC/run_fastqc.bat   \n",
            "  inflating: FastQC/fastqc_icon.ico  \n",
            "  inflating: FastQC/cisd-jhdf5.jar   \n",
            "  inflating: FastQC/jbzip2-0.9.jar   \n",
            "   creating: FastQC/uk/\n",
            "   creating: FastQC/uk/ac/\n",
            "   creating: FastQC/uk/ac/babraham/\n",
            "   creating: FastQC/uk/ac/babraham/FastQC/\n",
            "   creating: FastQC/uk/ac/babraham/FastQC/Graphs/\n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Graphs/LineGraph.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Graphs/BaseGroup.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Graphs/QualityBoxPlot.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Graphs/TileGraph.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/FastQCConfig.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/FastQCApplication$1.class  \n",
            "   creating: FastQC/uk/ac/babraham/FastQC/Results/\n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Results/ResultsPanel.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Results/ResultsPanel$ModuleRenderer.class  \n",
            "   creating: FastQC/uk/ac/babraham/FastQC/Utilities/\n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Utilities/NameFormatException.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Utilities/HotColdColourGradient.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Utilities/CasavaBasename.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Utilities/RGB.class  \n",
            "   creating: FastQC/uk/ac/babraham/FastQC/Utilities/ImageSaver/\n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Utilities/ImageSaver/SVGImageSaver.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Utilities/ImageSaver/SVGGenerator.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Utilities/ImageSaver/SVGGenerator$SVGGraphics.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Utilities/ImageToBase64.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Utilities/QualityCount.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Utilities/NanoporeBasename.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Utilities/MultiMemberGZIPInputStream.class  \n",
            "   creating: FastQC/uk/ac/babraham/FastQC/Sequence/\n",
            "   creating: FastQC/uk/ac/babraham/FastQC/Sequence/Contaminant/\n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Sequence/Contaminant/ContaminentFinder.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Sequence/Contaminant/Contaminant.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Sequence/Contaminant/ContaminantHit.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Sequence/FastQFile.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Sequence/Sequence.class  \n",
            "   creating: FastQC/uk/ac/babraham/FastQC/Sequence/QualityEncoding/\n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Sequence/QualityEncoding/PhredEncoding.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Sequence/Fast5File.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Sequence/BAMFile.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Sequence/SequenceFactory.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Sequence/SequenceFormatException.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Sequence/SequenceFileGroup.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Sequence/SequenceFile.class  \n",
            "   creating: FastQC/uk/ac/babraham/FastQC/Report/\n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Report/stylesheet.txt  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Report/HTMLReportArchive.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/FastQCMenuBar.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/FastQCApplication.class  \n",
            "   creating: FastQC/uk/ac/babraham/FastQC/Statistics/\n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Statistics/NormalDistribution.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Statistics/PearsonCorrelation.class  \n",
            "   creating: FastQC/uk/ac/babraham/FastQC/Analysis/\n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Analysis/AnalysisQueue.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Analysis/AnalysisListener.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Analysis/AnalysisRunner.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Analysis/OfflineRunner.class  \n",
            "   creating: FastQC/uk/ac/babraham/FastQC/Dialogs/\n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Dialogs/AboutDialog$1.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Dialogs/FastQCTitlePanel$SmoothJLabel.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Dialogs/WelcomePanel.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Dialogs/FastQCTitlePanel.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Dialogs/AboutDialog.class  \n",
            "   creating: FastQC/uk/ac/babraham/FastQC/FileFilters/\n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/FileFilters/FastQFileFilter.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/FileFilters/SequenceFileFilter.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/FileFilters/MappedBAMFileFilter.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/FileFilters/GobyFileFilter.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/FileFilters/CasavaFastQFileFilter.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/FileFilters/BAMFileFilter.class  \n",
            "   creating: FastQC/uk/ac/babraham/FastQC/Modules/\n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/KmerContent$Kmer.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/ModuleConfig.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/PerBaseQualityScores.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/OverRepresentedSeqs$OverrepresentedSeq.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/NContent.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/DuplicationLevel.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/SequenceLengthDistribution.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/BasicStats$ResultsTable.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/KmerContent$ResultsTable.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/PerBaseSequenceContent.class  \n",
            "   creating: FastQC/uk/ac/babraham/FastQC/Modules/GCModel/\n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/GCModel/GCModelValue.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/GCModel/GCModel.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/OverRepresentedSeqs$ResultsTable.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/AdapterContent.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/PerTileQualityScores.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/PerSequenceQualityScores.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/QCModule.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/KmerContent.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/AbstractQCModule.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/ModuleFactory.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/OverRepresentedSeqs.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/AdapterContent$Adapter.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/BasicStats.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/PerSequenceGCContent.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/AdapterContent$ResultsTable.class  \n",
            "   creating: FastQC/uk/ac/babraham/FastQC/Help/\n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Help/HelpPageDisplay.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Help/HelpIndexRoot$FileSorter.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Help/HelpDialog.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Help/HelpSearchPanel.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Help/HelpPageDisplay$HelpEditor.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Help/HelpIndexRoot.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Help/HelpPage.class  \n",
            "   creating: FastQC/uk/ac/babraham/FastQC/Resources/\n",
            " extracting: FastQC/uk/ac/babraham/FastQC/Resources/fastqc_icon_100.png  \n",
            " extracting: FastQC/uk/ac/babraham/FastQC/Resources/tick.png  \n",
            " extracting: FastQC/uk/ac/babraham/FastQC/Resources/error.png  \n",
            " extracting: FastQC/uk/ac/babraham/FastQC/Resources/fastqc_icon.png  \n",
            " extracting: FastQC/uk/ac/babraham/FastQC/Resources/warning.png  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Resources/fastqc_icon.svg  \n",
            "   creating: FastQC/Help/\n",
            "   creating: FastQC/Help/2 Basic Operations/\n",
            "   creating: FastQC/Help/2 Basic Operations/.svn/\n",
            "   creating: FastQC/Help/2 Basic Operations/.svn/tmp/\n",
            "   creating: FastQC/Help/2 Basic Operations/.svn/tmp/props/\n",
            "   creating: FastQC/Help/2 Basic Operations/.svn/props/\n",
            "   creating: FastQC/Help/2 Basic Operations/.svn/text-base/\n",
            "  inflating: FastQC/Help/2 Basic Operations/.svn/text-base/2.3 Saving a Report.html.svn-base  \n",
            "  inflating: FastQC/Help/2 Basic Operations/.svn/text-base/2.1 Opening a sequence file.html.svn-base  \n",
            "  inflating: FastQC/Help/2 Basic Operations/.svn/text-base/2.2 Evaluating Results.html.svn-base  \n",
            "  inflating: FastQC/Help/2 Basic Operations/.svn/entries  \n",
            "  inflating: FastQC/Help/2 Basic Operations/2.1 Opening a sequence file.html  \n",
            "  inflating: FastQC/Help/2 Basic Operations/2.3 Saving a Report.html  \n",
            "  inflating: FastQC/Help/2 Basic Operations/2.2 Evaluating Results.html  \n",
            "   creating: FastQC/Help/1 Introduction/\n",
            "  inflating: FastQC/Help/1 Introduction/1.1 What is FastQC.html  \n",
            "   creating: FastQC/Help/1 Introduction/.svn/\n",
            "   creating: FastQC/Help/1 Introduction/.svn/tmp/\n",
            "   creating: FastQC/Help/1 Introduction/.svn/tmp/props/\n",
            "   creating: FastQC/Help/1 Introduction/.svn/props/\n",
            "   creating: FastQC/Help/1 Introduction/.svn/text-base/\n",
            "  inflating: FastQC/Help/1 Introduction/.svn/text-base/1.1 What is FastQC.html.svn-base  \n",
            "  inflating: FastQC/Help/1 Introduction/.svn/entries  \n",
            "   creating: FastQC/Help/3 Analysis Modules/\n",
            "  inflating: FastQC/Help/3 Analysis Modules/per_base_quality.png  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/2 Per Base Sequence Quality.html  \n",
            "   creating: FastQC/Help/3 Analysis Modules/.svn/\n",
            "   creating: FastQC/Help/3 Analysis Modules/.svn/prop-base/\n",
            " extracting: FastQC/Help/3 Analysis Modules/.svn/prop-base/per_tile_quality.png.svn-base  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/.svn/prop-base/kmer_profiles.png.svn-base  \n",
            " extracting: FastQC/Help/3 Analysis Modules/.svn/prop-base/sequence_length_distribution.png.svn-base  \n",
            " extracting: FastQC/Help/3 Analysis Modules/.svn/prop-base/per_sequence_quality.png.svn-base  \n",
            " extracting: FastQC/Help/3 Analysis Modules/.svn/prop-base/per_base_n_content.png.svn-base  \n",
            " extracting: FastQC/Help/3 Analysis Modules/.svn/prop-base/per_base_gc_content.png.svn-base  \n",
            " extracting: FastQC/Help/3 Analysis Modules/.svn/prop-base/per_base_quality.png.svn-base  \n",
            " extracting: FastQC/Help/3 Analysis Modules/.svn/prop-base/per_sequence_gc_content.png.svn-base  \n",
            " extracting: FastQC/Help/3 Analysis Modules/.svn/prop-base/per_base_sequence_content.png.svn-base  \n",
            " extracting: FastQC/Help/3 Analysis Modules/.svn/prop-base/duplication_levels.png.svn-base  \n",
            "   creating: FastQC/Help/3 Analysis Modules/.svn/tmp/\n",
            "   creating: FastQC/Help/3 Analysis Modules/.svn/tmp/props/\n",
            "   creating: FastQC/Help/3 Analysis Modules/.svn/props/\n",
            "   creating: FastQC/Help/3 Analysis Modules/.svn/text-base/\n",
            "  inflating: FastQC/Help/3 Analysis Modules/.svn/text-base/8 Duplicate Sequences.html.svn-base  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/.svn/text-base/7 Sequence Length Distribution.html.svn-base  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/.svn/text-base/11 Kmer Content.html.svn-base  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/.svn/text-base/3 Per Sequence Quality Scores.html.svn-base  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/.svn/text-base/per_tile_quality.png.svn-base  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/.svn/text-base/12 Per Tile Sequence Quality.html.svn-base  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/.svn/text-base/kmer_profiles.png.svn-base  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/.svn/text-base/sequence_length_distribution.png.svn-base  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/.svn/text-base/4 Per Base Sequence Content.html.svn-base  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/.svn/text-base/9 Overrepresented Sequences.html.svn-base  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/.svn/text-base/2 Per Base Sequence Quality.html.svn-base  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/.svn/text-base/per_sequence_quality.png.svn-base  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/.svn/text-base/6 Per Base N Content.html.svn-base  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/.svn/text-base/per_base_n_content.png.svn-base  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/.svn/text-base/per_base_gc_content.png.svn-base  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/.svn/text-base/per_base_quality.png.svn-base  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/.svn/text-base/10 Adapter Content.html.svn-base  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/.svn/text-base/5 Per Sequence GC Content.html.svn-base  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/.svn/text-base/per_sequence_gc_content.png.svn-base  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/.svn/text-base/1 Basic Statistics.html.svn-base  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/.svn/text-base/per_base_sequence_content.png.svn-base  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/.svn/text-base/duplication_levels.png.svn-base  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/.svn/entries  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/8 Duplicate Sequences.html  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/6 Per Base N Content.html  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/sequence_length_distribution.png  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/per_base_n_content.png  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/11 Kmer Content.html  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/per_base_sequence_content.png  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/12 Per Tile Sequence Quality.html  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/per_sequence_quality.png  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/per_tile_quality.png  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/4 Per Base Sequence Content.html  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/9 Overrepresented Sequences.html  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/per_sequence_gc_content.png  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/3 Per Sequence Quality Scores.html  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/10 Adapter Content.html  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/1 Basic Statistics.html  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/kmer_profiles.png  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/5 Per Sequence GC Content.html  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/duplication_levels.png  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/7 Sequence Length Distribution.html  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/per_base_gc_content.png  \n",
            "   creating: FastQC/org/\n",
            "   creating: FastQC/org/apache/\n",
            "   creating: FastQC/org/apache/commons/\n",
            "   creating: FastQC/org/apache/commons/math3/\n",
            "   creating: FastQC/org/apache/commons/math3/special/\n",
            "  inflating: FastQC/org/apache/commons/math3/special/Beta$1.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/special/Erf.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/special/Gamma.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/special/Beta.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/special/Gamma$1.class  \n",
            "   creating: FastQC/org/apache/commons/math3/util/\n",
            "  inflating: FastQC/org/apache/commons/math3/util/Incrementor$1.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/util/Incrementor.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/util/ArithmeticUtils.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/util/ContinuedFraction.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/util/Precision.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/util/FastMath$ExpIntTable.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/util/MathUtils.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/util/Incrementor$MaxCountExceededCallback.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/util/FastMathLiteralArrays.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/util/FastMath$lnMant.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/util/FastMath$ExpFracTable.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/util/FastMathCalc.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/util/DoubleArray.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/util/FastMath.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/util/ResizableDoubleArray.class  \n",
            "   creating: FastQC/org/apache/commons/math3/distribution/\n",
            "  inflating: FastQC/org/apache/commons/math3/distribution/PascalDistribution.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/distribution/BinomialDistribution.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/distribution/NormalDistribution.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/distribution/CauchyDistribution.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/distribution/TDistribution.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/distribution/HypergeometricDistribution.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/distribution/GammaDistribution.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/distribution/PoissonDistribution.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/distribution/SaddlePointExpansion.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/distribution/FDistribution.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/distribution/BetaDistribution.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/distribution/AbstractRealDistribution$1.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/distribution/ZipfDistribution.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/distribution/IntegerDistribution.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/distribution/AbstractIntegerDistribution.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/distribution/RealDistribution.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/distribution/ChiSquaredDistribution.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/distribution/WeibullDistribution.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/distribution/AbstractRealDistribution.class  \n",
            "   creating: FastQC/org/apache/commons/math3/exception/\n",
            "  inflating: FastQC/org/apache/commons/math3/exception/MathInternalError.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/exception/NotStrictlyPositiveException.class  \n",
            "   creating: FastQC/org/apache/commons/math3/exception/util/\n",
            "  inflating: FastQC/org/apache/commons/math3/exception/util/ExceptionContextProvider.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/exception/util/ArgUtils.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/exception/util/ExceptionContext.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/exception/util/Localizable.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/exception/util/LocalizedFormats.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/exception/TooManyEvaluationsException.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/exception/MathIllegalStateException.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/exception/MathArithmeticException.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/exception/MathIllegalNumberException.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/exception/MathIllegalArgumentException.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/exception/OutOfRangeException.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/exception/NumberIsTooSmallException.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/exception/MaxCountExceededException.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/exception/NotPositiveException.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/exception/NoBracketingException.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/exception/NotFiniteNumberException.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/exception/NullArgumentException.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/exception/DimensionMismatchException.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/exception/ConvergenceException.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/exception/NumberIsTooLargeException.class  \n",
            "   creating: FastQC/org/apache/commons/math3/random/\n",
            "  inflating: FastQC/org/apache/commons/math3/random/Well19937c.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/random/AbstractWell.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/random/RandomGenerator.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/random/RandomDataImpl.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/random/RandomData.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/random/BitsStreamGenerator.class  \n",
            "   creating: FastQC/org/apache/commons/math3/analysis/\n",
            "  inflating: FastQC/org/apache/commons/math3/analysis/UnivariateFunction.class  \n",
            "   creating: FastQC/org/apache/commons/math3/analysis/solvers/\n",
            "  inflating: FastQC/org/apache/commons/math3/analysis/solvers/BaseAbstractUnivariateSolver.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/analysis/solvers/AllowedSolution.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/analysis/solvers/UnivariateSolverUtils.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/analysis/solvers/BrentSolver.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/analysis/solvers/UnivariateSolver.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/analysis/solvers/BaseUnivariateSolver.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/analysis/solvers/AbstractUnivariateSolver.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/analysis/solvers/BracketedUnivariateSolver.class  \n",
            "Skipping '/content/test.fastq' which didn't exist, or couldn't be read\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Update and install samtools\n",
        "!apt-get update\n",
        "!apt-get install -y samtools\n",
        "!samtools --version"
      ],
      "metadata": {
        "id": "nA2ABFrRSxsZ",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "9d244b41-9595-4183-f19a-6cefc9d175a4"
      },
      "execution_count": 3,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\r0% [Working]\r            \rGet:1 https://cli.github.com/packages stable InRelease [3,917 B]\n",
            "\r0% [Connecting to archive.ubuntu.com] [Waiting for headers] [Waiting for header\r0% [Connecting to archive.ubuntu.com] [Waiting for headers] [Waiting for header\r                                                                               \rGet:2 https://cloud.r-project.org/bin/linux/ubuntu jammy-cran40/ InRelease [3,632 B]\n",
            "\r0% [Connecting to archive.ubuntu.com] [Waiting for headers] [Connected to r2u.s\r                                                                               \rGet:3 https://developer.download.nvidia.com/compute/cuda/repos/ubuntu2204/x86_64  InRelease [1,581 B]\n",
            "Get:4 http://security.ubuntu.com/ubuntu jammy-security InRelease [129 kB]\n",
            "Get:5 https://r2u.stat.illinois.edu/ubuntu jammy InRelease [6,555 B]\n",
            "Hit:6 http://archive.ubuntu.com/ubuntu jammy InRelease\n",
            "Get:7 https://developer.download.nvidia.com/compute/cuda/repos/ubuntu2204/x86_64  Packages [2,045 kB]\n",
            "Get:8 http://archive.ubuntu.com/ubuntu jammy-updates InRelease [128 kB]\n",
            "Get:9 https://r2u.stat.illinois.edu/ubuntu jammy/main amd64 Packages [2,807 kB]\n",
            "Hit:10 https://ppa.launchpadcontent.net/deadsnakes/ppa/ubuntu jammy InRelease\n",
            "Get:11 http://security.ubuntu.com/ubuntu jammy-security/universe amd64 Packages [1,274 kB]\n",
            "Hit:12 https://ppa.launchpadcontent.net/graphics-drivers/ppa/ubuntu jammy InRelease\n",
            "Hit:13 https://ppa.launchpadcontent.net/ubuntugis/ppa/ubuntu jammy InRelease\n",
            "Get:14 https://r2u.stat.illinois.edu/ubuntu jammy/main all Packages [9,313 kB]\n",
            "Get:15 http://archive.ubuntu.com/ubuntu jammy-backports InRelease [127 kB]\n",
            "Get:16 http://security.ubuntu.com/ubuntu jammy-security/main amd64 Packages [3,371 kB]\n",
            "Get:17 http://archive.ubuntu.com/ubuntu jammy-updates/main amd64 Packages [3,695 kB]\n",
            "Get:18 http://archive.ubuntu.com/ubuntu jammy-updates/universe amd64 Packages [1,580 kB]\n",
            "Get:19 http://archive.ubuntu.com/ubuntu jammy-backports/universe amd64 Packages [35.2 kB]\n",
            "Fetched 24.5 MB in 3s (7,339 kB/s)\n",
            "Reading package lists... Done\n",
            "W: Skipping acquire of configured file 'main/source/Sources' as repository 'https://r2u.stat.illinois.edu/ubuntu jammy InRelease' does not seem to provide it (sources.list entry misspelt?)\n",
            "Reading package lists... Done\n",
            "Building dependency tree... Done\n",
            "Reading state information... Done\n",
            "The following additional packages will be installed:\n",
            "  libhts3 libhtscodecs2\n",
            "Suggested packages:\n",
            "  cwltool\n",
            "The following NEW packages will be installed:\n",
            "  libhts3 libhtscodecs2 samtools\n",
            "0 upgraded, 3 newly installed, 0 to remove and 50 not upgraded.\n",
            "Need to get 963 kB of archives.\n",
            "After this operation, 2,270 kB of additional disk space will be used.\n",
            "Get:1 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libhtscodecs2 amd64 1.1.1-3 [53.2 kB]\n",
            "Get:2 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libhts3 amd64 1.13+ds-2build1 [390 kB]\n",
            "Get:3 http://archive.ubuntu.com/ubuntu jammy/universe amd64 samtools amd64 1.13-4 [520 kB]\n",
            "Fetched 963 kB in 1s (748 kB/s)\n",
            "Selecting previously unselected package libhtscodecs2:amd64.\n",
            "(Reading database ... 126441 files and directories currently installed.)\n",
            "Preparing to unpack .../libhtscodecs2_1.1.1-3_amd64.deb ...\n",
            "Unpacking libhtscodecs2:amd64 (1.1.1-3) ...\n",
            "Selecting previously unselected package libhts3:amd64.\n",
            "Preparing to unpack .../libhts3_1.13+ds-2build1_amd64.deb ...\n",
            "Unpacking libhts3:amd64 (1.13+ds-2build1) ...\n",
            "Selecting previously unselected package samtools.\n",
            "Preparing to unpack .../samtools_1.13-4_amd64.deb ...\n",
            "Unpacking samtools (1.13-4) ...\n",
            "Setting up libhtscodecs2:amd64 (1.1.1-3) ...\n",
            "Setting up libhts3:amd64 (1.13+ds-2build1) ...\n",
            "Setting up samtools (1.13-4) ...\n",
            "Processing triggers for man-db (2.10.2-1) ...\n",
            "Processing triggers for libc-bin (2.35-0ubuntu3.8) ...\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtcm.so.1 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbb.so.12 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbbind_2_5.so.3 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbmalloc.so.2 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libur_adapter_opencl.so.0 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbbind.so.3 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libumf.so.0 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libhwloc.so.15 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbmalloc_proxy.so.2 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbbind_2_0.so.3 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libur_adapter_level_zero_v2.so.0 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtcm_debug.so.1 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libur_adapter_level_zero.so.0 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libur_loader.so.0 is not a symbolic link\n",
            "\n",
            "samtools 1.13\n",
            "Using htslib 1.13+ds\n",
            "Copyright (C) 2021 Genome Research Ltd.\n",
            "\n",
            "Samtools compilation details:\n",
            "    Features:       build=configure curses=yes \n",
            "    CC:             gcc\n",
            "    CPPFLAGS:       -frelease  -Wdate-time -D_FORTIFY_SOURCE=2\n",
            "    CFLAGS:         -g -O2 -ffile-prefix-map=�BUILDPATH�=. -flto=auto -ffat-lto-objects -fstack-protector-strong -Wformat -Werror=format-security\n",
            "    LDFLAGS:        -Wl,-Bsymbolic-functions -flto=auto -Wl,-z,relro -Wl,-z,now\n",
            "    HTSDIR:         \n",
            "    LIBS:           \n",
            "    CURSES_LIB:     -lcurses\n",
            "\n",
            "HTSlib compilation details:\n",
            "    Features:       build=configure plugins=yes, plugin-path=/usr/local/lib/htslib:/usr/local/libexec/htslib:: libcurl=yes S3=yes GCS=yes libdeflate=yes lzma=yes bzip2=yes htscodecs=1.1.1\n",
            "    CC:             gcc\n",
            "    CPPFLAGS:       -I. -DSAMTOOLS=1 -Wdate-time -D_FORTIFY_SOURCE=2\n",
            "    CFLAGS:         -g -O2 -ffile-prefix-map=/build/htslib-TQtOKr/htslib-1.13+ds=. -flto=auto -ffat-lto-objects -fstack-protector-strong -Wformat -Werror=format-security -flto -fvisibility=hidden -flto -fvisibility=hidden\n",
            "    LDFLAGS:        -Wl,-Bsymbolic-functions -flto=auto -Wl,-z,relro -Wl,-z,now -Wl,-flto -fvisibility=hidden -Wl,-flto -fvisibility=hidden\n",
            "\n",
            "HTSlib URL scheme handlers present:\n",
            "    built-in:\t preload, data, file\n",
            "    crypt4gh-needed:\t crypt4gh\n",
            "    mem:\t mem\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!apt-get update\n",
        "!apt-get install -y subread\n",
        "!featureCounts -v"
      ],
      "metadata": {
        "id": "MPiAd5qRS4rp",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "5cf07de4-2d91-458a-e5fa-b71c56eee311"
      },
      "execution_count": 4,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\r0% [Working]\r            \rGet:1 https://cli.github.com/packages stable InRelease [3,917 B]\n",
            "\r0% [Connecting to archive.ubuntu.com (185.125.190.36)] [Connecting to security.\r0% [Connecting to archive.ubuntu.com (185.125.190.36)] [Connecting to security.\r                                                                               \rHit:2 https://developer.download.nvidia.com/compute/cuda/repos/ubuntu2204/x86_64  InRelease\n",
            "\r0% [Waiting for headers] [Waiting for headers] [Waiting for headers] [Connected\r                                                                               \rHit:3 https://cloud.r-project.org/bin/linux/ubuntu jammy-cran40/ InRelease\n",
            "\r0% [Waiting for headers] [Waiting for headers] [Connected to r2u.stat.illinois.\r0% [Waiting for headers] [Waiting for headers] [Connected to r2u.stat.illinois.\r0% [Waiting for headers] [Waiting for headers] [Waiting for headers] [Connected\r                                                                               \rHit:4 http://archive.ubuntu.com/ubuntu jammy InRelease\n",
            "Hit:5 http://security.ubuntu.com/ubuntu jammy-security InRelease\n",
            "Hit:6 https://r2u.stat.illinois.edu/ubuntu jammy InRelease\n",
            "Hit:7 http://archive.ubuntu.com/ubuntu jammy-updates InRelease\n",
            "Hit:8 http://archive.ubuntu.com/ubuntu jammy-backports InRelease\n",
            "Hit:9 https://ppa.launchpadcontent.net/deadsnakes/ppa/ubuntu jammy InRelease\n",
            "Hit:10 https://ppa.launchpadcontent.net/graphics-drivers/ppa/ubuntu jammy InRelease\n",
            "Hit:11 https://ppa.launchpadcontent.net/ubuntugis/ppa/ubuntu jammy InRelease\n",
            "Fetched 3,917 B in 1s (2,651 B/s)\n",
            "Reading package lists... Done\n",
            "W: Skipping acquire of configured file 'main/source/Sources' as repository 'https://r2u.stat.illinois.edu/ubuntu jammy InRelease' does not seem to provide it (sources.list entry misspelt?)\n",
            "Reading package lists... Done\n",
            "Building dependency tree... Done\n",
            "Reading state information... Done\n",
            "The following NEW packages will be installed:\n",
            "  subread\n",
            "0 upgraded, 1 newly installed, 0 to remove and 50 not upgraded.\n",
            "Need to get 565 kB of archives.\n",
            "After this operation, 1,928 kB of additional disk space will be used.\n",
            "Get:1 http://archive.ubuntu.com/ubuntu jammy/universe amd64 subread amd64 2.0.3+dfsg-1 [565 kB]\n",
            "Fetched 565 kB in 1s (543 kB/s)\n",
            "Selecting previously unselected package subread.\n",
            "(Reading database ... 126540 files and directories currently installed.)\n",
            "Preparing to unpack .../subread_2.0.3+dfsg-1_amd64.deb ...\n",
            "Unpacking subread (2.0.3+dfsg-1) ...\n",
            "Setting up subread (2.0.3+dfsg-1) ...\n",
            "Processing triggers for man-db (2.10.2-1) ...\n",
            "\n",
            "featureCounts v2.0.3\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Download Trimmomatic\n",
        "!wget http://www.usadellab.org/cms/uploads/supplementary/Trimmomatic/Trimmomatic-0.39.zip\n",
        "!unzip Trimmomatic-0.39.zip\n",
        "!java -jar Trimmomatic-0.39/trimmomatic-0.39.jar"
      ],
      "metadata": {
        "id": "0PAdRWBJS_QR",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "437f01a4-44b4-4d7a-fba8-f79c78e54ea5"
      },
      "execution_count": 5,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "--2025-09-30 19:30:48--  http://www.usadellab.org/cms/uploads/supplementary/Trimmomatic/Trimmomatic-0.39.zip\n",
            "Resolving www.usadellab.org (www.usadellab.org)... 80.83.125.111\n",
            "Connecting to www.usadellab.org (www.usadellab.org)|80.83.125.111|:80... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 133596 (130K) [application/zip]\n",
            "Saving to: ‘Trimmomatic-0.39.zip’\n",
            "\n",
            "Trimmomatic-0.39.zi 100%[===================>] 130.46K   259KB/s    in 0.5s    \n",
            "\n",
            "2025-09-30 19:30:49 (259 KB/s) - ‘Trimmomatic-0.39.zip’ saved [133596/133596]\n",
            "\n",
            "Archive:  Trimmomatic-0.39.zip\n",
            "   creating: Trimmomatic-0.39/\n",
            "  inflating: Trimmomatic-0.39/LICENSE  \n",
            "  inflating: Trimmomatic-0.39/trimmomatic-0.39.jar  \n",
            "   creating: Trimmomatic-0.39/adapters/\n",
            "  inflating: Trimmomatic-0.39/adapters/NexteraPE-PE.fa  \n",
            "  inflating: Trimmomatic-0.39/adapters/TruSeq2-PE.fa  \n",
            "  inflating: Trimmomatic-0.39/adapters/TruSeq2-SE.fa  \n",
            "  inflating: Trimmomatic-0.39/adapters/TruSeq3-PE-2.fa  \n",
            "  inflating: Trimmomatic-0.39/adapters/TruSeq3-PE.fa  \n",
            "  inflating: Trimmomatic-0.39/adapters/TruSeq3-SE.fa  \n",
            "Usage: \n",
            "       PE [-version] [-threads <threads>] [-phred33|-phred64] [-trimlog <trimLogFile>] [-summary <statsSummaryFile>] [-quiet] [-validatePairs] [-basein <inputBase> | <inputFile1> <inputFile2>] [-baseout <outputBase> | <outputFile1P> <outputFile1U> <outputFile2P> <outputFile2U>] <trimmer1>...\n",
            "   or: \n",
            "       SE [-version] [-threads <threads>] [-phred33|-phred64] [-trimlog <trimLogFile>] [-summary <statsSummaryFile>] [-quiet] <inputFile> <outputFile> <trimmer1>...\n",
            "   or: \n",
            "       -version\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# e) Downloading the chr22 of the human reference genome hg38 (USCSC) and also the GTF file that contains the genes and their location (annotation file).\n",
        "\n",
        "After downloading, we are going to build a reference genome for STAR.\n",
        "\n",
        "In this step we need to include a parameter: --sjdbOverhang $read_length -1, which is related to the kmers STAR is going to use. for this number, we used read length -1"
      ],
      "metadata": {
        "id": "kLpTTu8jTZPs"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!wget http://hgdownload.soe.ucsc.edu/goldenPath/hg38/chromosomes/chr22.fa.gz\n",
        "!gunzip chr22.fa.gz\n",
        "!wget https://hgdownload.soe.ucsc.edu/goldenPath/hg38/bigZips/genes/hg38.ensGene.gtf.gz\n",
        "!gunzip hg38.ensGene.gtf.gz"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "pcUoPakRM8Sk",
        "outputId": "de204626-f56a-4237-a1e5-b98377c2daad"
      },
      "execution_count": 6,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "--2025-09-30 19:30:55--  http://hgdownload.soe.ucsc.edu/goldenPath/hg38/chromosomes/chr22.fa.gz\n",
            "Resolving hgdownload.soe.ucsc.edu (hgdownload.soe.ucsc.edu)... 128.114.119.163\n",
            "Connecting to hgdownload.soe.ucsc.edu (hgdownload.soe.ucsc.edu)|128.114.119.163|:80... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 12255678 (12M) [application/x-gzip]\n",
            "Saving to: ‘chr22.fa.gz’\n",
            "\n",
            "chr22.fa.gz         100%[===================>]  11.69M  51.4MB/s    in 0.2s    \n",
            "\n",
            "2025-09-30 19:30:55 (51.4 MB/s) - ‘chr22.fa.gz’ saved [12255678/12255678]\n",
            "\n",
            "--2025-09-30 19:30:56--  https://hgdownload.soe.ucsc.edu/goldenPath/hg38/bigZips/genes/hg38.ensGene.gtf.gz\n",
            "Resolving hgdownload.soe.ucsc.edu (hgdownload.soe.ucsc.edu)... 128.114.119.163\n",
            "Connecting to hgdownload.soe.ucsc.edu (hgdownload.soe.ucsc.edu)|128.114.119.163|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 27802050 (27M) [application/x-gzip]\n",
            "Saving to: ‘hg38.ensGene.gtf.gz’\n",
            "\n",
            "hg38.ensGene.gtf.gz 100%[===================>]  26.51M  91.1MB/s    in 0.3s    \n",
            "\n",
            "2025-09-30 19:30:56 (91.1 MB/s) - ‘hg38.ensGene.gtf.gz’ saved [27802050/27802050]\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!./STAR/bin/Linux_x86_64_static/STAR --runThreadN 6 --runMode genomeGenerate --genomeDir Gencode_STAR_Index --genomeFastaFiles chr22.fa --sjdbGTFfile hg38.ensGene.gtf --sjdbOverhang 49 --genomeSAindexNbases 11"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "tYKPi6ycM8_f",
        "outputId": "12e11b79-7d12-40d2-8b64-4965dbde8edc"
      },
      "execution_count": 7,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\t./STAR/bin/Linux_x86_64_static/STAR --runThreadN 6 --runMode genomeGenerate --genomeDir Gencode_STAR_Index --genomeFastaFiles chr22.fa --sjdbGTFfile hg38.ensGene.gtf --sjdbOverhang 49 --genomeSAindexNbases 11\n",
            "\tSTAR version: 2.7.11b   compiled: 2024-01-25T16:12:02-05:00 :/home/dobin/data/STAR/STARcode/STAR.master/source\n",
            "Sep 30 19:31:04 ..... started STAR run\n",
            "Sep 30 19:31:04 ... starting to generate Genome files\n",
            "Sep 30 19:31:05 ..... processing annotations GTF\n",
            "Sep 30 19:31:12 ... starting to sort Suffix Array. This may take a long time...\n",
            "Sep 30 19:31:13 ... sorting Suffix Array chunks and saving them to disk...\n",
            "Sep 30 19:32:01 ... loading chunks from disk, packing SA...\n",
            "Sep 30 19:32:03 ... finished generating suffix array\n",
            "Sep 30 19:32:03 ... generating Suffix Array index\n",
            "Sep 30 19:32:07 ... completed Suffix Array index\n",
            "Sep 30 19:32:07 ..... inserting junctions into the genome indices\n",
            "Sep 30 19:32:12 ... writing Genome to disk ...\n",
            "Sep 30 19:32:12 ... writing Suffix Array to disk ...\n",
            "Sep 30 19:32:13 ... writing SAindex to disk\n",
            "Sep 30 19:32:13 ..... finished successfully\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Now we need to get the raw data for our analysis. Usually we will just download it from SRA but because we don't want to overwhelm google collab, we will use a file I filtered for chr22 reads in the github repository"
      ],
      "metadata": {
        "id": "ty2dxiVJLb9b"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!git clone https://github.com/almejiaga/RNA-seq-bootcamp-day-2.git\n",
        "!gunzip RNA-seq-bootcamp-day-2/Exercises/data/unique_chr22.fastq.gz"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "aCnMX-5SN-Ww",
        "outputId": "56b53eca-feb2-4eef-ca08-896806838b89"
      },
      "execution_count": 8,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Cloning into 'RNA-seq-bootcamp-day-2'...\n",
            "remote: Enumerating objects: 21, done.\u001b[K\n",
            "remote: Counting objects: 100% (21/21), done.\u001b[K\n",
            "remote: Compressing objects: 100% (16/16), done.\u001b[K\n",
            "remote: Total 21 (delta 1), reused 12 (delta 0), pack-reused 0 (from 0)\u001b[K\n",
            "Receiving objects: 100% (21/21), 22.50 MiB | 33.94 MiB/s, done.\n",
            "Resolving deltas: 100% (1/1), done.\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Running FASTQC to get the QC metrics we saw in the lecture"
      ],
      "metadata": {
        "id": "1biHgXhDLjpS"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!FastQC/fastqc RNA-seq-bootcamp-day-2/Exercises/data/unique_chr22.fastq"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "qpdtANCHHKXz",
        "outputId": "79e38b55-6c06-48f4-9c4f-8df39cd0b641"
      },
      "execution_count": 10,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "null\n",
            "Started analysis of unique_chr22.fastq\n",
            "Approx 5% complete for unique_chr22.fastq\n",
            "Approx 10% complete for unique_chr22.fastq\n",
            "Approx 15% complete for unique_chr22.fastq\n",
            "Approx 20% complete for unique_chr22.fastq\n",
            "Approx 25% complete for unique_chr22.fastq\n",
            "Approx 30% complete for unique_chr22.fastq\n",
            "Approx 35% complete for unique_chr22.fastq\n",
            "Approx 40% complete for unique_chr22.fastq\n",
            "Approx 45% complete for unique_chr22.fastq\n",
            "Approx 50% complete for unique_chr22.fastq\n",
            "Approx 55% complete for unique_chr22.fastq\n",
            "Approx 60% complete for unique_chr22.fastq\n",
            "Approx 65% complete for unique_chr22.fastq\n",
            "Approx 70% complete for unique_chr22.fastq\n",
            "Approx 75% complete for unique_chr22.fastq\n",
            "Approx 80% complete for unique_chr22.fastq\n",
            "Approx 85% complete for unique_chr22.fastq\n",
            "Approx 90% complete for unique_chr22.fastq\n",
            "Approx 95% complete for unique_chr22.fastq\n",
            "Analysis complete for unique_chr22.fastq\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!ls FastQC/fastqc RNA-seq-bootcamp-day-2/Exercises/data/"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "XcbFgIFLHbt7",
        "outputId": "86228700-0295-46a0-83d2-9f12955e603b"
      },
      "execution_count": 12,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "FastQC/fastqc\n",
            "\n",
            "RNA-seq-bootcamp-day-2/Exercises/data/:\n",
            "data.txt  unique_chr22.fastq  unique_chr22_fastqc.html\tunique_chr22_fastqc.zip\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Trimming our raw reads to ensure the best quality for aligment"
      ],
      "metadata": {
        "id": "I2KeSsxqU0s1"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!java -jar Trimmomatic-0.39/trimmomatic-0.39.jar SE \\\n",
        "  -threads 4 \\\n",
        "  RNA-seq-bootcamp-day-2/Exercises/data/unique_chr22.fastq \\\n",
        "  R1_trimmed.fastq \\\n",
        "  ILLUMINACLIP:Trimmomatic-0.39/adapters/TruSeq3-SE.fa:2:30:10 \\\n",
        "  LEADING:3 TRAILING:3 SLIDINGWINDOW:4:15 MINLEN:36\n"
      ],
      "metadata": {
        "id": "oNQDpcnFU075",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "007b3bb2-5172-4e81-bca2-1563e67fab6e"
      },
      "execution_count": 19,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "TrimmomaticSE: Started with arguments:\n",
            " -threads 4 RNA-seq-bootcamp-day-2/Exercises/data/unique_chr22.fastq R1_trimmed.fastq ILLUMINACLIP:Trimmomatic-0.39/adapters/TruSeq3-SE.fa:2:30:10 LEADING:3 TRAILING:3 SLIDINGWINDOW:4:15 MINLEN:36\n",
            "Using Long Clipping Sequence: 'AGATCGGAAGAGCGTCGTGTAGGGAAAGAGTGTA'\n",
            "Using Long Clipping Sequence: 'AGATCGGAAGAGCACACGTCTGAACTCCAGTCAC'\n",
            "ILLUMINACLIP: Using 0 prefix pairs, 2 forward/reverse sequences, 0 forward only sequences, 0 reverse only sequences\n",
            "Quality encoding detected as phred33\n",
            "Input Reads: 674468 Surviving: 672201 (99.66%) Dropped: 2267 (0.34%)\n",
            "TrimmomaticSE: Completed successfully\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Running FASTQC after trimming to see if the QC metric have improved or not"
      ],
      "metadata": {
        "id": "l7fgojlbL_A5"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!FastQC/fastqc R1_trimmed.fastq"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "xsppeo7DIuf5",
        "outputId": "c8be22b8-9798-40f2-9e26-54b97681f4ac"
      },
      "execution_count": 21,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "null\n",
            "Started analysis of R1_trimmed.fastq\n",
            "Approx 5% complete for R1_trimmed.fastq\n",
            "Approx 10% complete for R1_trimmed.fastq\n",
            "Approx 15% complete for R1_trimmed.fastq\n",
            "Approx 20% complete for R1_trimmed.fastq\n",
            "Approx 25% complete for R1_trimmed.fastq\n",
            "Approx 30% complete for R1_trimmed.fastq\n",
            "Approx 35% complete for R1_trimmed.fastq\n",
            "Approx 40% complete for R1_trimmed.fastq\n",
            "Approx 45% complete for R1_trimmed.fastq\n",
            "Approx 50% complete for R1_trimmed.fastq\n",
            "Approx 55% complete for R1_trimmed.fastq\n",
            "Approx 60% complete for R1_trimmed.fastq\n",
            "Approx 65% complete for R1_trimmed.fastq\n",
            "Approx 70% complete for R1_trimmed.fastq\n",
            "Approx 75% complete for R1_trimmed.fastq\n",
            "Approx 80% complete for R1_trimmed.fastq\n",
            "Approx 85% complete for R1_trimmed.fastq\n",
            "Approx 90% complete for R1_trimmed.fastq\n",
            "Approx 95% complete for R1_trimmed.fastq\n",
            "Analysis complete for R1_trimmed.fastq\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Alignment\n",
        "\n",
        "After trimming the FASTQ and keeping only the high quality reads, we are ready to align to the human genome. we need to specify the folder that contains the index for the reference. There are a lot of other parameters associated with the alignment that we play with."
      ],
      "metadata": {
        "id": "ddTLRbI5UAN5"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!./STAR/bin/Linux_x86_64_static/STAR --genomeDir Gencode_STAR_Index --runThreadN 1 --readFilesIn R1_trimmed.fastq --outSAMtype BAM SortedByCoordinate --limitBAMsortRAM 17179869184 --outFilterType BySJout --outFilterMultimapNmax 20 --alignSJoverhangMin 8 --alignSJDBoverhangMin 1 --outFilterMismatchNmax 999 --outFilterMismatchNoverLmax 0.04 --alignIntronMin 20 --alignIntronMax 1000000 --seedSearchStartLmax 30 --outFileNamePrefix 1_mapped_chr22"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "gzUg5ulNO8_g",
        "outputId": "6f61dd43-515b-421f-ce7e-e7b13060b824"
      },
      "execution_count": 22,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\t./STAR/bin/Linux_x86_64_static/STAR --genomeDir Gencode_STAR_Index --runThreadN 1 --readFilesIn R1_trimmed.fastq --outSAMtype BAM SortedByCoordinate --limitBAMsortRAM 17179869184 --outFilterType BySJout --outFilterMultimapNmax 20 --alignSJoverhangMin 8 --alignSJDBoverhangMin 1 --outFilterMismatchNmax 999 --outFilterMismatchNoverLmax 0.04 --alignIntronMin 20 --alignIntronMax 1000000 --seedSearchStartLmax 30 --outFileNamePrefix 1_mapped_chr22\n",
            "\tSTAR version: 2.7.11b   compiled: 2024-01-25T16:12:02-05:00 :/home/dobin/data/STAR/STARcode/STAR.master/source\n",
            "Sep 30 19:40:12 ..... started STAR run\n",
            "Sep 30 19:40:12 ..... loading genome\n",
            "Sep 30 19:40:13 ..... started mapping\n",
            "Sep 30 19:42:37 ..... finished mapping\n",
            "Sep 30 19:42:37 ..... started sorting BAM\n",
            "Sep 30 19:42:40 ..... finished successfully\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Indexing and filtering out reads with low mapping quality scores"
      ],
      "metadata": {
        "id": "AHpHjoLcMJ9a"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "\n",
        "!samtools index 1_mapped_chr22Aligned.sortedByCoord.out.bam\n",
        "!samtools view -F 2820 -q 30 -@ 4 -b 1_mapped_chr22Aligned.sortedByCoord.out.bam > 1_filtered.bam"
      ],
      "metadata": {
        "id": "hLZsKX5QKBub"
      },
      "execution_count": 26,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Generaring the count matrix\n",
        "\n"
      ],
      "metadata": {
        "id": "tMH3I7iGUbiz"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!featureCounts -T 4 -t exon -g gene_id -a hg38.ensGene.gtf -o one_sample.txt 1_filtered.bam"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "jBzjB3ALP7Vo",
        "outputId": "3e203e43-26fe-43e1-dd7d-141f243c68ca"
      },
      "execution_count": 27,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\n",
            "       \u001b[44;37m ========== \u001b[0m\u001b[36m    _____ _    _ ____  _____  ______          _____  \n",
            "       \u001b[44;37m =====      \u001b[0m\u001b[36m   / ____| |  | |  _ \\|  __ \\|  ____|   /\\   |  __ \\ \n",
            "       \u001b[44;37m   =====    \u001b[0m\u001b[36m  | (___ | |  | | |_) | |__) | |__     /  \\  | |  | |\n",
            "       \u001b[44;37m     ====   \u001b[0m\u001b[36m   \\___ \\| |  | |  _ <|  _  /|  __|   / /\\ \\ | |  | |\n",
            "       \u001b[44;37m       ==== \u001b[0m\u001b[36m   ____) | |__| | |_) | | \\ \\| |____ / ____ \\| |__| |\n",
            "       \u001b[44;37m ========== \u001b[0m\u001b[36m  |_____/ \\____/|____/|_|  \\_\\______/_/    \\_\\_____/\u001b[0m\n",
            "\t  v2.0.3\n",
            "\n",
            "//==========================\u001b[36m featureCounts setting \u001b[0m===========================\\\\\n",
            "||  \u001b[0m                                                                          ||\n",
            "||             Input files : \u001b[36m1 BAM file  \u001b[0m \u001b[0m                                    ||\n",
            "||  \u001b[0m                                                                          ||\n",
            "||                           \u001b[36m1_filtered.bam\u001b[0m \u001b[0m                                  ||\n",
            "||  \u001b[0m                                                                          ||\n",
            "||             Output file : \u001b[36mone_sample.txt\u001b[0m \u001b[0m                                  ||\n",
            "||                 Summary : \u001b[36mone_sample.txt.summary\u001b[0m \u001b[0m                          ||\n",
            "||              Paired-end : \u001b[36mno\u001b[0m \u001b[0m                                              ||\n",
            "||        Count read pairs : \u001b[36mno\u001b[0m \u001b[0m                                              ||\n",
            "||              Annotation : \u001b[36mhg38.ensGene.gtf (GTF)\u001b[0m \u001b[0m                          ||\n",
            "||      Dir for temp files : \u001b[36m./\u001b[0m \u001b[0m                                              ||\n",
            "||  \u001b[0m                                                                          ||\n",
            "||                 Threads : \u001b[36m4\u001b[0m \u001b[0m                                               ||\n",
            "||                   Level : \u001b[36mmeta-feature level\u001b[0m \u001b[0m                              ||\n",
            "||      Multimapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
            "|| Multi-overlapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
            "||   Min overlapping bases : \u001b[36m1\u001b[0m \u001b[0m                                               ||\n",
            "||  \u001b[0m                                                                          ||\n",
            "\\\\============================================================================//\n",
            "\n",
            "//=================================\u001b[36m Running \u001b[0m==================================\\\\\n",
            "||  \u001b[0m                                                                          ||\n",
            "|| Load annotation file hg38.ensGene.gtf \u001b[0m... \u001b[0m                                 ||\n",
            "||    Features : \u001b[36m1232593\u001b[0m \u001b[0m                                                     ||\n",
            "||    Meta-features : \u001b[36m64252\u001b[0m \u001b[0m                                                  ||\n",
            "||    Chromosomes/contigs : \u001b[36m270\u001b[0m \u001b[0m                                              ||\n",
            "||  \u001b[0m                                                                          ||\n",
            "|| Process BAM file 1_filtered.bam... \u001b[0m                                        ||\n",
            "||    Single-end reads are included. \u001b[0m                                         ||\n",
            "||    Total alignments : \u001b[36m671380\u001b[0m \u001b[0m                                              ||\n",
            "||    Successfully assigned alignments : \u001b[36m456267 (68.0%)\u001b[0m \u001b[0m                      ||\n",
            "||    Running time : \u001b[36m0.01 minutes\u001b[0m \u001b[0m                                            ||\n",
            "||  \u001b[0m                                                                          ||\n",
            "|| Write the final count table. \u001b[0m                                              ||\n",
            "|| Write the read assignment summary. \u001b[0m                                        ||\n",
            "||  \u001b[0m                                                                          ||\n",
            "|| Summary of counting results can be found in file \"one_sample.txt.summary\" \u001b[0m ||\n",
            "||  \u001b[0m                                                                          ||\n",
            "\\\\============================================================================//\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Loading the count matrix in python and checking the top hits\n",
        "\n",
        "If we search the top genes, they should be related to constitutive genes or macrophages-related genes"
      ],
      "metadata": {
        "id": "RsPmheWYUaDJ"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "import pandas as pd\n",
        "\n",
        "# Read the featureCounts output, skip header lines\n",
        "df = pd.read_csv(\"one_sample.txt\", sep=\"\\t\", comment=\"#\")\n",
        "\n",
        "# Get the BAM column name (it usually matches your BAM filename)\n",
        "bam_col = \"1_filtered.bam\"\n",
        "\n",
        "# Sort by counts in descending order and show top 10\n",
        "top10 = df.sort_values(by=bam_col, ascending=False).head(10)\n",
        "\n",
        "top10"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 383
        },
        "id": "ENnPOq7CSe9D",
        "outputId": "be927d84-b703-416c-ef7e-2cfe5e0582a5"
      },
      "execution_count": 29,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "                Geneid                                                Chr  \\\n",
              "38710  ENSG00000128340  chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr2...   \n",
              "38858  ENSG00000213857                                              chr22   \n",
              "37804  ENSG00000241838                                              chr22   \n",
              "38673  ENSG00000100345  chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr2...   \n",
              "38727  ENSG00000100097  chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr2...   \n",
              "38809  ENSG00000100316  chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr2...   \n",
              "38575  ENSG00000232346                                              chr22   \n",
              "38511  ENSG00000225774                                              chr22   \n",
              "38995  ENSG00000100364  chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr2...   \n",
              "37860  ENSG00000177663  chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr2...   \n",
              "\n",
              "                                                   Start  \\\n",
              "38710  37225261;37225852;37225966;37226671;37226671;3...   \n",
              "38858                                           41074180   \n",
              "37804                                           15721486   \n",
              "38673  36281281;36284093;36284162;36284214;36284403;3...   \n",
              "38727  37675608;37675652;37675657;37675682;37675688;3...   \n",
              "38809  39312882;39312884;39312884;39312884;39312884;3...   \n",
              "38575                                           32039490   \n",
              "38511                                           30542536   \n",
              "38995  45190338;45195828;45195828;45195828;45196038;4...   \n",
              "37860  17084954;17084959;17084959;17085057;17097062;1...   \n",
              "\n",
              "                                                     End  \\\n",
              "38710  37226039;37226039;37226039;37226803;37226803;3...   \n",
              "38858                                           41075239   \n",
              "37804                                           15722608   \n",
              "38673  36282785;36284265;36284307;36284511;36284511;3...   \n",
              "38727  37675711;37675711;37675711;37675711;37675711;3...   \n",
              "38809  39312984;39313310;39312984;39312984;39312984;3...   \n",
              "38575                                           32039896   \n",
              "38511                                           30544305   \n",
              "38995  45197216;45197216;45197216;45197216;45197216;4...   \n",
              "37860  17085229;17085229;17085229;17085229;17097509;1...   \n",
              "\n",
              "                                                  Strand  Length  \\\n",
              "38710  -;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;...    2532   \n",
              "38858                                                  -    1060   \n",
              "37804                                                  +    1123   \n",
              "38673  -;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;...    8959   \n",
              "38727  +;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+    1099   \n",
              "38809  -;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;...    4787   \n",
              "38575                                                  +     407   \n",
              "38511                                                  +    1770   \n",
              "38995  -;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;...   10019   \n",
              "37860  +;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;...    9036   \n",
              "\n",
              "       1_filtered.bam  \n",
              "38710           27818  \n",
              "38858           23101  \n",
              "37804           21142  \n",
              "38673           19757  \n",
              "38727           13966  \n",
              "38809            9528  \n",
              "38575            9274  \n",
              "38511            6603  \n",
              "38995            6178  \n",
              "37860            5974  "
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-55f0ee78-a3aa-4ed9-9bd5-1dca210bc15f\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Geneid</th>\n",
              "      <th>Chr</th>\n",
              "      <th>Start</th>\n",
              "      <th>End</th>\n",
              "      <th>Strand</th>\n",
              "      <th>Length</th>\n",
              "      <th>1_filtered.bam</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>38710</th>\n",
              "      <td>ENSG00000128340</td>\n",
              "      <td>chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr2...</td>\n",
              "      <td>37225261;37225852;37225966;37226671;37226671;3...</td>\n",
              "      <td>37226039;37226039;37226039;37226803;37226803;3...</td>\n",
              "      <td>-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;...</td>\n",
              "      <td>2532</td>\n",
              "      <td>27818</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>38858</th>\n",
              "      <td>ENSG00000213857</td>\n",
              "      <td>chr22</td>\n",
              "      <td>41074180</td>\n",
              "      <td>41075239</td>\n",
              "      <td>-</td>\n",
              "      <td>1060</td>\n",
              "      <td>23101</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>37804</th>\n",
              "      <td>ENSG00000241838</td>\n",
              "      <td>chr22</td>\n",
              "      <td>15721486</td>\n",
              "      <td>15722608</td>\n",
              "      <td>+</td>\n",
              "      <td>1123</td>\n",
              "      <td>21142</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>38673</th>\n",
              "      <td>ENSG00000100345</td>\n",
              "      <td>chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr2...</td>\n",
              "      <td>36281281;36284093;36284162;36284214;36284403;3...</td>\n",
              "      <td>36282785;36284265;36284307;36284511;36284511;3...</td>\n",
              "      <td>-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;...</td>\n",
              "      <td>8959</td>\n",
              "      <td>19757</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>38727</th>\n",
              "      <td>ENSG00000100097</td>\n",
              "      <td>chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr2...</td>\n",
              "      <td>37675608;37675652;37675657;37675682;37675688;3...</td>\n",
              "      <td>37675711;37675711;37675711;37675711;37675711;3...</td>\n",
              "      <td>+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+</td>\n",
              "      <td>1099</td>\n",
              "      <td>13966</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>38809</th>\n",
              "      <td>ENSG00000100316</td>\n",
              "      <td>chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr2...</td>\n",
              "      <td>39312882;39312884;39312884;39312884;39312884;3...</td>\n",
              "      <td>39312984;39313310;39312984;39312984;39312984;3...</td>\n",
              "      <td>-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;...</td>\n",
              "      <td>4787</td>\n",
              "      <td>9528</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>38575</th>\n",
              "      <td>ENSG00000232346</td>\n",
              "      <td>chr22</td>\n",
              "      <td>32039490</td>\n",
              "      <td>32039896</td>\n",
              "      <td>+</td>\n",
              "      <td>407</td>\n",
              "      <td>9274</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>38511</th>\n",
              "      <td>ENSG00000225774</td>\n",
              "      <td>chr22</td>\n",
              "      <td>30542536</td>\n",
              "      <td>30544305</td>\n",
              "      <td>+</td>\n",
              "      <td>1770</td>\n",
              "      <td>6603</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>38995</th>\n",
              "      <td>ENSG00000100364</td>\n",
              "      <td>chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr2...</td>\n",
              "      <td>45190338;45195828;45195828;45195828;45196038;4...</td>\n",
              "      <td>45197216;45197216;45197216;45197216;45197216;4...</td>\n",
              "      <td>-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;...</td>\n",
              "      <td>10019</td>\n",
              "      <td>6178</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>37860</th>\n",
              "      <td>ENSG00000177663</td>\n",
              "      <td>chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr2...</td>\n",
              "      <td>17084954;17084959;17084959;17085057;17097062;1...</td>\n",
              "      <td>17085229;17085229;17085229;17085229;17097509;1...</td>\n",
              "      <td>+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;...</td>\n",
              "      <td>9036</td>\n",
              "      <td>5974</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-55f0ee78-a3aa-4ed9-9bd5-1dca210bc15f')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-55f0ee78-a3aa-4ed9-9bd5-1dca210bc15f button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-55f0ee78-a3aa-4ed9-9bd5-1dca210bc15f');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "    <div id=\"df-82e84556-1177-4318-a6a4-f12ee06cb667\">\n",
              "      <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-82e84556-1177-4318-a6a4-f12ee06cb667')\"\n",
              "                title=\"Suggest charts\"\n",
              "                style=\"display:none;\">\n",
              "\n",
              "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "     width=\"24px\">\n",
              "    <g>\n",
              "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "    </g>\n",
              "</svg>\n",
              "      </button>\n",
              "\n",
              "<style>\n",
              "  .colab-df-quickchart {\n",
              "      --bg-color: #E8F0FE;\n",
              "      --fill-color: #1967D2;\n",
              "      --hover-bg-color: #E2EBFA;\n",
              "      --hover-fill-color: #174EA6;\n",
              "      --disabled-fill-color: #AAA;\n",
              "      --disabled-bg-color: #DDD;\n",
              "  }\n",
              "\n",
              "  [theme=dark] .colab-df-quickchart {\n",
              "      --bg-color: #3B4455;\n",
              "      --fill-color: #D2E3FC;\n",
              "      --hover-bg-color: #434B5C;\n",
              "      --hover-fill-color: #FFFFFF;\n",
              "      --disabled-bg-color: #3B4455;\n",
              "      --disabled-fill-color: #666;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart {\n",
              "    background-color: var(--bg-color);\n",
              "    border: none;\n",
              "    border-radius: 50%;\n",
              "    cursor: pointer;\n",
              "    display: none;\n",
              "    fill: var(--fill-color);\n",
              "    height: 32px;\n",
              "    padding: 0;\n",
              "    width: 32px;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart:hover {\n",
              "    background-color: var(--hover-bg-color);\n",
              "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "    fill: var(--button-hover-fill-color);\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart-complete:disabled,\n",
              "  .colab-df-quickchart-complete:disabled:hover {\n",
              "    background-color: var(--disabled-bg-color);\n",
              "    fill: var(--disabled-fill-color);\n",
              "    box-shadow: none;\n",
              "  }\n",
              "\n",
              "  .colab-df-spinner {\n",
              "    border: 2px solid var(--fill-color);\n",
              "    border-color: transparent;\n",
              "    border-bottom-color: var(--fill-color);\n",
              "    animation:\n",
              "      spin 1s steps(1) infinite;\n",
              "  }\n",
              "\n",
              "  @keyframes spin {\n",
              "    0% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "      border-left-color: var(--fill-color);\n",
              "    }\n",
              "    20% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    30% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    40% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    60% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    80% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "    90% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "  }\n",
              "</style>\n",
              "\n",
              "      <script>\n",
              "        async function quickchart(key) {\n",
              "          const quickchartButtonEl =\n",
              "            document.querySelector('#' + key + ' button');\n",
              "          quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
              "          quickchartButtonEl.classList.add('colab-df-spinner');\n",
              "          try {\n",
              "            const charts = await google.colab.kernel.invokeFunction(\n",
              "                'suggestCharts', [key], {});\n",
              "          } catch (error) {\n",
              "            console.error('Error during call to suggestCharts:', error);\n",
              "          }\n",
              "          quickchartButtonEl.classList.remove('colab-df-spinner');\n",
              "          quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
              "        }\n",
              "        (() => {\n",
              "          let quickchartButtonEl =\n",
              "            document.querySelector('#df-82e84556-1177-4318-a6a4-f12ee06cb667 button');\n",
              "          quickchartButtonEl.style.display =\n",
              "            google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "        })();\n",
              "      </script>\n",
              "    </div>\n",
              "\n",
              "  <div id=\"id_fd46f523-b1c8-4026-87ff-8ccb4939aff3\">\n",
              "    <style>\n",
              "      .colab-df-generate {\n",
              "        background-color: #E8F0FE;\n",
              "        border: none;\n",
              "        border-radius: 50%;\n",
              "        cursor: pointer;\n",
              "        display: none;\n",
              "        fill: #1967D2;\n",
              "        height: 32px;\n",
              "        padding: 0 0 0 0;\n",
              "        width: 32px;\n",
              "      }\n",
              "\n",
              "      .colab-df-generate:hover {\n",
              "        background-color: #E2EBFA;\n",
              "        box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "        fill: #174EA6;\n",
              "      }\n",
              "\n",
              "      [theme=dark] .colab-df-generate {\n",
              "        background-color: #3B4455;\n",
              "        fill: #D2E3FC;\n",
              "      }\n",
              "\n",
              "      [theme=dark] .colab-df-generate:hover {\n",
              "        background-color: #434B5C;\n",
              "        box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "        filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "        fill: #FFFFFF;\n",
              "      }\n",
              "    </style>\n",
              "    <button class=\"colab-df-generate\" onclick=\"generateWithVariable('top10')\"\n",
              "            title=\"Generate code using this dataframe.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M7,19H8.4L18.45,9,17,7.55,7,17.6ZM5,21V16.75L18.45,3.32a2,2,0,0,1,2.83,0l1.4,1.43a1.91,1.91,0,0,1,.58,1.4,1.91,1.91,0,0,1-.58,1.4L9.25,21ZM18.45,9,17,7.55Zm-12,3A5.31,5.31,0,0,0,4.9,8.1,5.31,5.31,0,0,0,1,6.5,5.31,5.31,0,0,0,4.9,4.9,5.31,5.31,0,0,0,6.5,1,5.31,5.31,0,0,0,8.1,4.9,5.31,5.31,0,0,0,12,6.5,5.46,5.46,0,0,0,6.5,12Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "    <script>\n",
              "      (() => {\n",
              "      const buttonEl =\n",
              "        document.querySelector('#id_fd46f523-b1c8-4026-87ff-8ccb4939aff3 button.colab-df-generate');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      buttonEl.onclick = () => {\n",
              "        google.colab.notebook.generateWithVariable('top10');\n",
              "      }\n",
              "      })();\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "dataframe",
              "variable_name": "top10",
              "summary": "{\n  \"name\": \"top10\",\n  \"rows\": 10,\n  \"fields\": [\n    {\n      \"column\": \"Geneid\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 10,\n        \"samples\": [\n          \"ENSG00000100364\",\n          \"ENSG00000213857\",\n          \"ENSG00000100316\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Chr\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 7,\n        \"samples\": [\n          \"chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr22\",\n          \"chr22\",\n          \"chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr22;chr22\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Start\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 10,\n        \"samples\": [\n          \"45190338;45195828;45195828;45195828;45196038;45196702;45196749;45196964;45197106;45197790;45197790;45197790;45197790;45197790;45197790;45197790;45197790;45197790;45199873;45199873;45199873;45199873;45199873;45199873;45199873;45199873;45202563;45202594;45202990;45202990;45202990;45202990;45202990;45203131;45203845;45203845;45203845;45203845;45203845;45203845;45205217;45205217;45205217;45205217;45205217;45205217;45205221;45205223;45205260;45205630;45205630;45205630;45205630;45205630;45205630;45205630;45205630;45205630;45205793;45205793;45205793;45205793;45205793;45205793;45205793;45205793;45205793;45205884;45207573;45211334;45211335;45211956;45211956;45211956;45211956;45211956;45211956;45211956;45211956;45211956;45211956;45211956;45212265;45212265;45213321;45226152;45226780;45235249;45236170;45240640;45240640\",\n          \"41074180\",\n          \"39312882;39312884;39312884;39312884;39312884;39312887;39313191;39313191;39313191;39313191;39313191;39313634;39313634;39313634;39313634;39313707;39313931;39314107;39314107;39314107;39314107;39314150;39314686;39314686;39314686;39314686;39314686;39314686;39314686;39314700;39315362;39315369;39315369;39315369;39315369;39315369;39315369;39315369;39315369;39315552;39316706;39316706;39316706;39316706;39316706;39316706;39316706;39316706;39317051;39317396;39317461;39317461;39317461;39317461;39317461;39317461;39317461;39317855;39318112;39318400;39318400;39318400;39318400;39318400;39318400;39318400;39318400;39318400;39318400;39319028;39319372;39319390;39319390;39319390;39319499;39319595;39319595;39319595;39319595;39319727\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"End\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 10,\n        \"samples\": [\n          \"45197216;45197216;45197216;45197216;45197216;45197216;45197216;45197216;45197216;45197948;45197948;45198033;45197948;45197948;45197948;45197948;45197948;45197948;45200035;45200035;45200209;45200035;45200035;45200035;45200035;45200035;45202715;45202684;45203184;45203184;45203184;45203184;45203184;45203184;45203985;45203985;45203985;45203985;45203985;45203985;45205318;45205318;45205318;45205318;45205318;45205318;45205318;45205318;45205318;45205707;45205707;45205707;45205707;45205707;45205641;45205707;45205707;45205707;45205912;45205912;45205912;45205912;45205912;45205849;45205912;45205912;45205912;45205912;45207772;45211432;45211432;45212107;45212107;45212445;45212107;45212107;45212081;45212107;45212457;45212107;45212107;45212107;45212417;45212488;45213499;45226383;45226903;45235413;45236277;45240742;45240769\",\n          \"41075239\",\n          \"39312984;39313310;39312984;39312984;39312984;39312984;39314208;39313705;39313310;39313310;39313310;39313729;39313729;39313729;39313729;39313729;39314208;39314208;39314208;39314208;39314208;39314408;39314846;39314846;39314965;39314846;39314846;39314846;39314846;39314846;39315555;39315555;39315429;39315486;39315555;39315555;39315555;39315555;39315555;39315555;39317021;39316841;39316841;39317068;39316841;39316841;39317021;39316841;39317629;39317629;39317629;39317629;39317629;39317629;39317629;39317629;39317629;39318592;39318592;39318592;39318592;39318592;39318592;39318592;39318592;39318592;39318592;39318592;39318614;39319111;39319625;39319577;39319592;39319623;39319595;39319649;39319623;39319623;39319771;39320389\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Strand\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 8,\n        \"samples\": [\n          \"-\",\n          \"-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-\",\n          \"-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-;-\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Length\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 3831,\n        \"min\": 407,\n        \"max\": 10019,\n        \"num_unique_values\": 10,\n        \"samples\": [\n          10019,\n          1060,\n          4787\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"1_filtered.bam\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 8023,\n        \"min\": 5974,\n        \"max\": 27818,\n        \"num_unique_values\": 10,\n        \"samples\": [\n          6178,\n          23101,\n          9528\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    }\n  ]\n}"
            }
          },
          "metadata": {},
          "execution_count": 29
        }
      ]
    }
  ]
}